subsys/net/ip/net_if.c

```diff
@@ -555,6 +561,90 @@ static void dad_timeout(struct k_work *work)
 		 */
 		net_ipv6_nbr_rm(iface, &ifaddr->address.in6_addr);
 	}
+
+	ifaddr->dad.timer_timeout = 0;
+
+	sys_slist_find_and_remove(&active_dad_timers, &ifaddr->dad.node);
+}
+
+static bool dad_check_timeout(s64_t start, u32_t time, s64_t timeout)
+{
+	start += time;
+	start = abs(start);
+
+	if (start > timeout) {
+		return false;
+	}
+
+	return true;
+}
+
+static bool dad_timedout(struct net_if_addr *ifaddr, s64_t timeout)
+{
+	return dad_check_timeout(ifaddr->dad.timer_start,
+				 ifaddr->dad.timer_timeout,
+				 timeout);
+}
+
+static u32_t dad_manage_timeouts(struct net_if_addr *ifaddr, s64_t timeout)
+{
+	s32_t next_timeout;
+
+	if (dad_timedout(ifaddr, timeout)) {
+		dad_expired(ifaddr);
+		return UINT32_MAX;
+	}
+
+	next_timeout = timeout - (ifaddr->dad.timer_start +
+				  ifaddr->dad.timer_timeout);
+	return abs(next_timeout);
+}
+
+static void dad_timeout(struct k_work *work)
+{
+	u32_t timeout_update = UINT32_MAX - 1;
+	s64_t timeout = k_uptime_get();
+	struct net_if_addr *current, *next;
+
+	ARG_UNUSED(work);
+
+	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&active_dad_timers,
+					  current, next, dad.node) {
+		u32_t next_timeout;
+
+		next_timeout = dad_manage_timeouts(current, timeout);
+		if (next_timeout < timeout_update) {
+			timeout_update = next_timeout;
+		}
+	}
+
+	if (timeout_update < (UINT32_MAX - 1)) {
```

Comment:
Apparently, here should be check for UINT32_MAX, it's the value which has a special meaning in dad_manage_timeouts() retval.

Reply:
Can you please elaborate how the entire processing works then?

Reply:
The timer_manage_timeout() will return UINT32_MAX if there was a timeout. The initial value is set to UINT32_MAX-1 so that we will know if there are any timers that should be triggered. The timeout_update will eventually contain the lowest timeout value that is found in the list. If there was none, then timeout_update will contain either UINT32_MAX or UINT32_MAX-1 value and no work needs to be triggered.

Reply:
Thanks for the detailed description, just what I was looking for! But my [rubber ducky](https://en.wikipedia.org/wiki/Rubber_duck_debugging) sees it differently:

Indeed, timer_manage_timeout() returns either UINT32_MAX to signify that timer already expired, or any other value (i.e. in the range 0..UINT32_MAX-1) to signify delay until the timer expiration. I.e. conceptually, it returns Union(0..UINT32_MAX-1, MAGIC_VALUE=UINT32_MAX).

Now it's caller, dad_timeout(), should also either deal with magic value of UINT32_MAX, or with a value in range 0..UINT32_MAX-1. But instead, it introduces another magic value, UINT32_MAX-1. Interpretation of that value appears to be the same as for UINT32_MAX, but there's a bug - if timer_manage_timeout() returns UINT32_MAX-1, which from its PoV is a normal timer delay, then dad_timeout() will interpret it as a magic value. But the main problem is that there's no need for another magic value at all, so only UINT32_MAX should be used throughout.

Reply:
Yes, I noticed it too, the -1 magic is gone already in the latest version.